### PR TITLE
BUG: loffset not applied when using resample with agg() (GH13218)

### LINKED ISSUE #13218: loffset doesn't work in resample if used with agg()
#### Code Sample, a copy-pastable example if possible

```
# Create DF
df = pd.DataFrame(np.random.rand(5,2), columns=list('AB'), index=pd.date_range('2010-01-01 09:00:00', periods=5, freq='s'))

print(df)
                                       A                    B
2010-01-01 09:00:00             0.283113             0.559642
2010-01-01 09:00:01             0.754942             0.621557
2010-01-01 09:00:02             0.102002             0.892100
2010-01-01 09:00:03             0.885400             0.524359
2010-01-01 09:00:04             0.324761             0.706758

#Resample, loffset works fine using mean(), is offset by 2 hours.
print(df.resample('2s', loffset='2h').mean())
                                       A                    B
2010-01-01 11:00:00             0.519028             0.590599
2010-01-01 11:00:02             0.493701             0.708229
2010-01-01 11:00:04             0.324761             0.706758


# Resample with agg(), loffset doesn't work anymore.
print(df.resample('2s', loffset='2h').agg(dict(A='sum', B='mean')))
                                      A                    B
2010-01-01 09:00:00             1.038055             0.590599
2010-01-01 09:00:02             0.987402             0.708229
2010-01-01 09:00:04             0.324761             0.706758

```
#### Expected Output

I'd expect the same as with mean(), this is how resample worked in the past with resample(how=...).

But maybe I also misunderstood something about the change. If so please enlighten me.

Since I need this to keep things going, I use the following workaround for now, please let me know if this is the way to go or if there's a more efficient way:

```
print(df.resample('2s', loffset='2h').agg(dict(A='sum', B='mean')).tshift(2, 'h'))
```
## INSTALLED VERSIONS

commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Darwin
OS-release: 15.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: de_DE.UTF-8

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.2
setuptools: 20.3
Cython: 0.23.4
numpy: 1.10.4
scipy: 0.17.1
statsmodels: 0.6.1
xarray: None
IPython: 4.2.0
sphinx: 1.3.5
patsy: 0.4.0
dateutil: 2.5.1
pytz: 2016.2
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.5.2
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.4
lxml: 3.6.0
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.39.0
pandas_datareader: 0.2.1
In [92]:

### PR DESCRIPTION
**Pull request by wcwagner**
- [x] closes #13218 
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tseries/tests/test_resample.py [hunk 1]
```diff
@@ -1118,6 +1118,26 @@ def test_resample_loffset_count(self):
...
 
+    def test_resample_loffset_agg(self):
+        # GH 13218
+        start_time = '1/1/2000 00:00:00'
+        rng = date_range(start_time, periods=50, freq='S')
+        vals = np.random.randn(50)
+        df = DataFrame({'A': vals, 'B': vals}, index=rng)
+
+        result = df.resample('10S', loffset='2h').agg({'A': 'count',
+                                                       'B': 'mean'})
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: can u test for a series aggregation as well

Comment (sinhrks) [inline pandas/tseries/tests/test_resample.py]: can u add ``PeriodIndex`` also.

Comment (wcwagner) [inline pandas/tseries/tests/test_resample.py]: @jreback @sinhrks I was way too naive with the fix, it ends up applying loffset twice when not passing a dict to agg(). I will try to get a better fix along with more tests tonight.

File: pandas/tseries/resample.py [hunk 2]
```diff
@@ -310,6 +310,9 @@ def aggregate(self, arg, *args, **kwargs):
                                                *args,
                                                **kwargs)
 
+        if isinstance(arg, dict):
```
Comment (jreback) [inline pandas/tseries/resample.py]: maybe we should do this check *inside* ``_apply_loffset``

File: pandas/tseries/tests/test_resample.py [hunk 3]
```diff
@@ -1118,6 +1118,43 @@ def test_resample_loffset_count(self):
 
         assert_series_equal(result, expected)
 
+    def test_resample_loffset_agg(self):
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: I would ideally like to move this test (and the loffset aboev) to the ``Base`` class as that fully tests DTI,PI, and TD indexes methodically (if that's too dififcult, then put the test method in *each* class), not in a single class.

File: pandas/tseries/tests/test_resample.py [hunk 4]
```diff
@@ -1118,6 +1118,43 @@ def test_resample_loffset_count(self):
...
+
+        df = DataFrame({'A': vals, 'B': vals}, index=rng)
+
+        result = df.resample('10S', loffset='2h').agg({'A': 'count',
+                                                       'B': 'mean'})
+        expected = DataFrame({'A': 10, 'B': expected_means},
+                             index=expected_index)
+        assert_frame_equal(result, expected)
+
+        # now test agg() with df w/ PeriodIndex
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: e.g. this needs to go into the PeriodIndex Test class class. Need a test for TD as well..

**Comment by jreback** [comment]
in any event, pls add the test for timedelta index, but assert it fails (and put a TODO for the xref issue #7687 )

**Comment by wcwagner** [comment]
@jreback I'm not quite sure what you mean by put a TODO, could you please clarify? 

Anyways, doing the tests was kind of tricky, so please lmk if I went in the right direction.

**Comment by jreback** [comment]
what I mean is can leave the timedelta fixes for another PR to fix
but u can out the tests in for them (and since they will fail u can just assert that they fail and add a comment, a todo for fixing later)

**Comment by wcwagner** [comment]
@jreback for future reference, when I update my PRs should I @ someone with write access to let them know?  Or should I just wait, and check on them intermittently, rebasing when they get out of date?

thanks

File: pandas/tseries/tests/test_resample.py [hunk 5]
```diff
@@ -699,6 +699,81 @@ def test_resample_empty_dtypes(self):
                         # (ex: doing mean with dtype of np.object)
                         pass
 
+    def test_resample_loffset_count(self):
+        s = self.create_series()
+        df = s.to_frame('value')
+        result = df.resample('2D', loffset='2H').count()
+
+        # GH 13022, 7687 resample w/ TimedeltaIndex results in incorrect index
+        if isinstance(df.index, TimedeltaIndex):
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: no - put in the separate test classes as I indicated

**Comment by jreback** [comment]
you can ping me if u want me to look

I added some comments

**Comment by wcwagner** [comment]
@jreback split the tests into separate test classes

File: pandas/tseries/resample.py [hunk 6]
```diff
@@ -368,7 +370,7 @@ def _gotitem(self, key, ndim, subset=None):
             return grouped
 
     def _groupby_and_aggregate(self, how, grouper=None, *args, **kwargs):
-        """ revaluate the obj with a groupby aggregation """
+        """ reevaluate the obj with a groupby aggregation """
```
Comment (jreback) [inline pandas/tseries/resample.py]: ``re-evaluate``

File: pandas/tseries/resample.py [hunk 7]
```diff
@@ -395,8 +397,8 @@ def _groupby_and_aggregate(self, how, grouper=None, *args, **kwargs):
 
         return self._wrap_result(result)
 
-    def _apply_loffset(self, result):
-        """if loffset if set, offset the result index"""
+    def _apply_loffset(self, result, arg=None):
```
Comment (jreback) [inline pandas/tseries/resample.py]: this is very confusing, why do we need ``arg``? pls add a ``Parameters`` section

File: pandas/tseries/tests/test_resample.py [hunk 8]
```diff
@@ -1098,25 +1098,35 @@ def test_resample_loffset(self):
...
+                                       freq='2D',
+                                       periods=len(df.index) / 2)
+        expected_index = expected_index + timedelta(hours=2)
+        expected = DataFrame({'value': 2},
+                             index=expected_index)
+        assert_frame_equal(result, expected)
 
-        assert_series_equal(result, expected)
+    def test_resample_loffset_agg(self):
+        # GH 13218
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: there need to go in the sub-classes. ``TestPeriodIndex`` and such.

File: pandas/tseries/resample.py [hunk 9]
```diff
@@ -395,8 +397,19 @@ def _groupby_and_aggregate(self, how, grouper=None, *args, **kwargs):
...
+        """
+        if loffset is set, offset the result index
+
+        Parameters
+        ----------
+        result : Series or DataFrame
+            the result of resample
+        arg : string or dict, optional
+            only passed in when agg() is used w/ resample
+            if string is passed to agg, loffset is already applied
```
Comment (jreback) [inline pandas/tseries/resample.py]: this is VERY confusing. pls explain.

**Comment by wcwagner** [comment]
@jreback I think I explained the use of ``arg`` sufficiently. 

I also didn't realize that ``loffset`` was broken when passing in a list to agg, e.g. ```df.resample('2d', loffset='2h').agg(['mean']).```, so I added a fix and put in the corresponding tests.

**Comment by wcwagner** [comment]
@jreback when you have time can you take a look at this, thanks

File: pandas/tseries/resample.py [hunk 10]
```diff
@@ -395,8 +397,23 @@ def _groupby_and_aggregate(self, how, grouper=None, *args, **kwargs):
...
-        """if loffset if set, offset the result index"""
+    def _apply_loffset(self, result, arg=None):
+        """
+        if loffset is set, offset the result index
+
+        Parameters
+        ----------
+        result : Series or DataFrame
+            the result of resample
+        arg : string, list, or dict, optional
```
Comment (jreback) [inline pandas/tseries/resample.py]: this needs to be refactored to not do this. It is unlike anything else in pandas to pass an argument to determine state.

**Comment by wcwagner** [comment]
@jreback I've tried, but can't think of any more alternative solutions. I went back to what I had originally, with the logic in the ``aggregate()`` function. 

if this is insufficient, please lmk if you want me to close the PR

**Comment by jreback** [comment]
well, then reopen if you have a better way.

**Comment by wcwagner** [comment]
@jreback I don't wanna create a dupe, so is there any way you can reopen this?

**Pull request closed without merging**